File: iconics-java/src/main/java/com/open/iconics/view/IconicsImageView.java
package com.open.iconics.view;

import android.content.Context;
import android.util.AttributeSet;

import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;
import androidx.appcompat.widget.AppCompatImageView;

import com.open.iconics.IconicsDrawable;
import com.open.iconics.internal.IconicsView;
import com.open.iconics.internal.IconicsViewsAttrsApplier;
import com.open.iconics.internal.IconicsViewsUtils;

import static androidx.annotation.RestrictTo.Scope.LIBRARY_GROUP;

public class IconicsImageView extends AppCompatImageView implements IconicsView {

    public IconicsImageView(Context context) {
        this(context, null);
    }

    public IconicsImageView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public IconicsImageView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        if (!isInEditMode()) {
            initialize(context, attrs, defStyle);
        }
    }

    @Override
    @RestrictTo(LIBRARY_GROUP)
    public void initialize(Context context, AttributeSet attrs, int defStyle) {
        //set the scale type for this view
        setScaleType(ScaleType.CENTER_INSIDE);

        applyAttr(context, attrs, defStyle);
    }

    @Override
    @RestrictTo(LIBRARY_GROUP)
    public void applyAttr(Context context, AttributeSet attrs, int defStyle) {
        setIcon(IconicsViewsAttrsApplier.getIconicsImageViewDrawable(context, attrs));
    }

    public void setIcon(@Nullable IconicsDrawable icon) {
        setImageDrawable(IconicsViewsUtils.checkAnimation(icon, this));
    }

    @Nullable
    public IconicsDrawable getIcon() {
        if (getDrawable() instanceof IconicsDrawable) {
            return ((IconicsDrawable) getDrawable());
        } else {
            return null;
        }
    }
}

File: iconics-kotlin/src/main/java/com/open/iconics/view/IconicsImageView.kt
package com.open.iconics.view

import android.content.Context
import android.util.AttributeSet
import android.widget.ImageView
import androidx.appcompat.widget.AppCompatImageView
import com.open.iconics.IconicsDrawable
import com.open.iconics.animation.tryToEnableIconicsAnimation
import com.open.iconics.internal.IconicsViewsAttrsApplier

open class IconicsImageView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : AppCompatImageView(context, attrs, defStyle) {

    var icon: IconicsDrawable?
        get() = drawable as? IconicsDrawable
        set(icon) = setImageDrawable(tryToEnableIconicsAnimation(icon))

    init {
        //set the scale type for this view
        scaleType = ImageView.ScaleType.CENTER_INSIDE

        icon = IconicsViewsAttrsApplier.getIconicsImageViewDrawable(context, attrs)
    }
}
